### PR TITLE
kbuildbarn: start formatting symlink urls

### PR DESCRIPTION
**Pull request by aaahrens**
What this does:

Generates out a generic LL of symlinks out of a NamedSetOfFiles protobuff message. Each node of the LL contains a source and destination symlink, formatted for bb_clientd cas and scratch directory.

### CODE REVIEW

File: lib/kbuildbarn/protoparse.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	DefaultBBClientdCasFileTemplate     = "cas/%s/blobs/file/%s"
+	DefaultBBClientdScratchFileTemplate = "scratch/%s/%s"
+)
+
+// BBClientDLink represents a single symlink of a file in the cas to a hard symlink in the scratch directory
+type BBClientDLink struct {
+	Src  string
+	Dest string
+	Next *BBClientDLink
+}
```
Comment (minor-fixes) [inline lib/kbuildbarn/protoparse.go:19]: Why a linked list over a slice?

File: lib/kbuildbarn/protoparse_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		},
+		{
+			Name: "tarball.tar", Digest: "digest3", Length: 777,
+		},
+	}}
+	baseName := "/foo/bar"
+	clusterName := "duster"
+	invocationPrefix := "invocation"
+	expected := map[string]string{
+		"/foo/bar/scratch/invocation/simple.txt":       "/foo/bar/cas/duster/blobs/file/digest0",
```
Comment (minor-fixes) [inline lib/kbuildbarn/protoparse_test.go:45]: Any advantage to generating relative symlinks?

File: lib/kbuildbarn/protoparse.go (hunk 1 above)
Comment (aaahrens) [reply lib/kbuildbarn/protoparse.go]: Change to slice

File: lib/kbuildbarn/protoparse_test.go (hunk 2 above)
Comment (aaahrens) [reply lib/kbuildbarn/protoparse_test.go:45]: Not really, we have 0 clarity as to where the commands to create symlinks gets run etc. absolute paths are much safer, especially for cleanup.

**Comment by ccontavalli** [comment]
If we use the scratch space under buildbarn, btw, we should use hard links rather than symlinks. They are entirely invisible to the application, and what scratch/ was created for.

File: lib/kbuildbarn/options.go [hunk 3]
```diff
@@ -69,13 +67,36 @@ func WithDirectoryUrlTemplate() Option {
...
 
-type templateArgsOption []interface{}
+type templateArgsOption [][]interface{}
 
 func (ta templateArgsOption) apply(opts *options) {
+	opts.TemplateArgs = ta[0]
+}
+
+func WithTemplateArgs(args []interface{}) Option {
+	return templateArgsOption{args}
```
Comment (minor-fixes) [inline lib/kbuildbarn/options.go:77]: Maybe one slice dimension can be dropped if this is `return templateArgsOption(args)`?

...and then just `opts.TemplateArgs = ta` above?

Comment (aaahrens) [reply lib/kbuildbarn/options.go:77]: So fun fact, if you wrap []interface{} with a custom type,  it is now [][]interface{} (functionally).

so if on line 76 you `fmt.Println(args)` => `[ foo bar ]`

and on line 96, you fmt.Println(ta) => `[ [ foo bar ] ] `


*but* you can declare TA as an []interface{} (where the interface{} is another []interface{}. This requires a type check.

*or* you can declare ta as a [][]interface{}, and the go compiler knows that you can unwrap.

File: lib/kbuildbarn/options.go [hunk 4]
```diff
@@ -69,13 +67,36 @@ func WithDirectoryUrlTemplate() Option {
 	return pathTemplateOption(DefaultDirectoryTemplate)
 }
 
-type templateArgsOption []interface{}
+type templateArgsOption [][]interface{}
```
[the remark this answers is not in the document]
Comment (aaahrens) [reply lib/kbuildbarn/options.go:70]: See below

File: lib/kbuildbarn/options.go (hunk 3 above)
Comment (aaahrens) [reply lib/kbuildbarn/options.go]: Changed to just convert the []string input to []interface on the flag itself, easy fix.

**Review by minor-fixes — APPROVED** [review]

**Pull request merged**